fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/cursors/throttled/ThrottledIteratorTest.java

```diff
@@ -0,0 +1,622 @@
+        List<CompletableFuture<Void>> futures = new ArrayList<>(numRecords);
+
+        final ItemHandler<Integer> itemHandler = (store, item, quotaManager) -> {
+            // First future hangs on, all others are immediately completed
+            CompletableFuture<Void> future = (item.get() == 0) ? new CompletableFuture<>() : CompletableFuture.completedFuture(null);
+            futures.add(future);
+            return future;
+        };
+        final Consumer<ThrottledRetryingIterator.QuotaManager> initNotification = quotaManager -> {
+            transactionStart.incrementAndGet();
+        };
+
+        ThrottledRetryingIterator<Integer> throttledIterator =
+                iteratorBuilder(numRecords, itemHandler, initNotification, null, -1, -1, -1, -1, null).build();
+        final CompletableFuture<Void> iterateAll;
+        try (FDBRecordContext context = openContext()) {
+            openSimpleRecordStore(context);
+            iterateAll = throttledIterator.iterateAll(recordStore.asBuilder());
+        }
+        // Closing the iterator before the first future completes
+        throttledIterator.close();
+        // Only first future in the list, none other was created since the first one didn't complete
+        assertThat(futures).hasSize(1);
+        assertThat(futures.get(0).isCompletedExceptionally()).isTrue();
+        assertThatThrownBy(() -> futures.get(0).get()).hasCauseInstanceOf(FDBDatabaseRunner.RunnerClosed.class);
+        // Overall status is failed because we can't runAsync() anymore
+        assertThatThrownBy(iterateAll::join).hasCauseInstanceOf(FDBDatabaseRunner.RunnerClosed.class);
+        // Only one transaction started (no retry), since the runner was closed
+        assertThat(transactionStart.get()).isOne();
+    }
+
+    private ThrottledRetryingIterator.Builder<Integer> iteratorBuilder(final int numRecords,
+                                                                       final ItemHandler<Integer> itemHandler,
+                                                                       final Consumer<ThrottledRetryingIterator.QuotaManager> initNotification,
+                                                                       final Consumer<ThrottledRetryingIterator.QuotaManager> successNotification,
+                                                                       final int maxPerSecLimit,
+                                                                       final int maxDeletedPerTransaction, final int numRetries,
+                                                                       final int transactionTimeMillis, final AtomicInteger limitRef) {
+
+        ThrottledRetryingIterator.Builder<Integer> throttledIterator = ThrottledRetryingIterator.builder(fdb, intCursor(numRecords, limitRef), itemHandler);
+
+        if (successNotification != null) {
+            throttledIterator.withTransactionSuccessNotification(successNotification);
```

Comment:
I like the way this `iteratorBuilder` allows testing of the unmodified default values.